ファイル関連のデータを取得するためのエンドポイントに関する説明を追加

a5ob7r opened the pull request:
# 概要

ファイルのバイナリや名前やIDなどの関連データを取得するためのエンドポイント群に関する詳細設計をOpenAPI v3.1.0に沿って記述した。

エラーレスポンスのパラメータは、OAuth 2.0やProblem DetailsのRFCを参考に記述した。

# 関連知識

- OpenAPI Specification v3.1.0 - https://spec.openapis.org/oas/v3.1.0
- JSON Schema - https://json-schema.org
- OAuth Parameters - https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml

# 補足

本プルリクは #7 の後続である。 #7 のマージ後にベースブランチをmainに変更する。

a5ob7r committed:
Add some description for the APIs to get a file and the properties

a5ob7r committed:
Change the endpoint name from `binary` to `content`

`binary` is a little bit formal expression.

a5ob7r committed:
Add a description about an alias GET operation

tk0miya commented on openapi.yaml:
```diff
@@ -520,13 +552,79 @@ paths:
   /files/{fileId}:
     get:
-      summary: ファイル名やその説明文、あるいはファイル自身などファイルの情報を返す
+      summary: /files/{fileId}/content のエイリアスである。
```
呼び出してくる相手はプログラムなのですが、クライアントA は `/file/1` を、クライアント B は `/file/1/content` をそれぞれ呼ぶのは問題ない (望ましい) ということですか?
サーバの実装が複雑になる一方で、誰かにとってエイリアスを提供するメリットがあまり思いつきませんでした。

a5ob7r replied:
ご指摘はもっともだと思っています。自分でもこの仕様でいいものかと考えていたので、だからこその言い訳としての説明文が書いてあります。

とはいえ、どうせだったら意見を聞いてみたいと思ったので、そのままPRに含めた次第です。

tk0miya replied:
なるほどね。意見を聞いてみたいというのはいい考えだと思う。

その場合、できれば PR のコメントなどで「意見ください」って書いてもらえるとありがたいかな。
というのは

* API ドキュメントの読者は「API クライアントの開発者」
* PR の読者は「チームメート」

なので、書くべきことが違うと思っています。

API クライアントの開発者は、このシステム、この API についてあまり詳しくない可能性があるので、

* どういう API の構成なのか
* (順序や条件がある場合は) どういう風に API を読んでほしいのか

などを説明するのがよいと思います。
このエイリアスについては、どちらを呼ぶべきなのかという情報を提供できるといいですよね。
「同等として扱っている」と説明されても、自分のアプリでどちらを呼ぶべきなのかって判断つきづらいと思いました。


※ 基本的には同じ機能を持った API はひとつしか提供しないのですが、パス変更などによって「廃止予定の旧 API」と「現在使うべき新 API」のふたつが提供されることがあったりします。その場合は旧 API の方に非推奨って書いたりしますね。

a5ob7r replied:
> このエイリアスについては、どちらを呼ぶべきなのかという情報を提供できるといいですよね。
> 「同等として扱っている」と説明されても、自分のアプリでどちらを呼ぶべきなのかって判断つきづらいと思いました。

なるほど。迷ってしまうというのは考えていませんでした。2つ同じAPIを用意したのは、説明にもあるように、論理的な解釈と直感による理解のどちらも正しいように感じたので、ユーザは自分の感覚にあう方を使えばいいと考えたからでした。

でもドキュメントを見れば使い方はわかるはずだから、わざわざ2つも用意する必要もないか。

tk0miya replied:
一度世に出した API は簡単に廃止することはできません。
呼び出している API クライアントができてしまうと、その対処を考える必要があります。
同じ機能を持った API を２つ用意してしまうと、今後のメンテナンスは両方の API に対してやっていく必要がありますよね。
たとえばファイルの属性が増えたとき、権限制御をつけたとき、ファイルの保存場所を変えたときなど。

メンテナンスコストが増えるのに対して、僕らが得られるメリットはなんでしょうか。
このシステムの売上が上がるでしょうか。システムの利用者の使い勝手が上がるでしょうか。

システム設計における (モデリングとしての) 綺麗さ、整合性といった、そういった面については否定するつもりはないものの、僕らが作っているのは稼働し続けるシステムです。統一された API、わかりやすい API といった面では利便性やユーザビリティが確保できるかもしれませんが (議論の余地はあると思いますが)、保守性や経済合理性といった面ではどうでしょうか。

ちょっと話を広げてしまいましたが、こうした側面もあるのを知っておくと良いかなーと思いました。


単に技術的な側面 (モデリングの側面) だけ取り上げた場合でも、同じ機能に対して２つの名前を与えるメリットは特にないと思います。プログラミング言語のような汎用的なものだとすると、いろんな名前を与えていろんな書き方を可能にするという選択もありかもしれませんが、API という限定された領域では、そういう自由を与えることについてメリットは特にないと思います。
※ ひとつのメソッドにいろんな呼び名がついている汎用的なものの例として Ruby がありますね。Ruby はいろんなエリアスを持たせている、やや稀有な例です。他の言語ではあまり見かけない気がします。

a5ob7r replied:
> プログラミング言語のような汎用的なものだとすると、いろんな名前を与えていろんな書き方を可能にするという選択もありかもしれませんが、

あー、まさにこちら寄りの考え方をしていました。コードを書くならこちらの方が楽しいだろうと考えていましたが、今後は実利面を考慮していけるようにします。

a5ob7r replied:
今回は、 `/files/{fileId}` へのGETリクエストの方を削除します。

tk0miya commented on openapi.yaml:
```diff
@@ -520,13 +552,79 @@ paths:
+      description: |
+        `/files/{fileId}` へのGETリクエストは、`/files{fileId}/content` へのGETリクエストのエイリアスになっている。
+
+        論理的にはファイルリソース(`/files/{fileId}`)は、ファイルの実態であるバイナリ(`/files{fileId}/content`)とそのメタデータ(`/files{fileId}/properties`)から構成されている。
+        また、実用上においても、例えばファイルのバイナリへの直リンクを利用したいといった状況が考えられるため、先の2つのデータが別々のURIで提供する。
+        一方で、直感的には、ファイルリソースは、ファイルの実態であるバイナリとも解釈可能である。
+        よって、ファイルリソースへのGETリクエストでは、ファイルの実態であるバイナリへのGETリクエストと同等であるとして扱っている。
```
雑談: アプリの仕様にもよるのですが、Rails アプリがファイルを返すのではなく、Rails アプリは S3 の URL を返し、ファイルを S3 から取得するという構造にすることが多そうです。
Rails はアプリケーションであるため、ファイルの配信をするのに向いていません。ファイルを配信するのに向いている仕組みを使うのが良いという考えです。
一般に、アプリケーションサーバのリソースのほうが高コストとされるので、なるべく Rails に計算させないという設計の考え方もあります。

たとえば、ファイルを配信している間、Rails サーバの接続上限が減ってしまう (コネクションが専有されてしまう) というのは問題ですよね。
10人が 1GB のファイルをダウンロードしようとすると、(適当ですが) 1分の間 Rails サーバへの接続が 10本となってしまいます。

a5ob7r committed:
Fix a typo of a key name

tk0miya committed:
Merge pull request #9 from sls-training/fix_key_name_typo

タイポ修正

a5ob7r committed:
Extract a schema for 422

a5ob7r committed:
Remove an unnecessary API alias

The API aliases, except for deprecations, just increases maintenance
costs and provide no benefits for us, especially for sales. So they
should be removed if no acceptable reasons.

tim-toga committed:
Merge pull request #10 from sls-training/extract_unprocessable_entity_errors

共通コンポーネントへのUnprocessable Entity用スキーマの抽出
